pkg/sos/objectstorage.go

```diff
@@ -0,0 +1,209 @@
+	resp, err := o.exoscaleClient.ListSosBucketsUsageWithResponse(ctx)
+	if err != nil {
+		return err
+	}
+
+	aggregatedBuckets := getAggregatedBuckets(ctx, *resp.JSON200.SosBucketsUsage, o.bucketDetails)
+	if len(aggregatedBuckets) == 0 {
+		log.Info("There are no bucket usage to be saved in the database")
+		return nil
+	}
+
+	o.aggregatedBuckets = aggregatedBuckets
+	return nil
+}
+
+func (o *ObjectStorage) fetchManagedBuckets(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Fetching buckets from cluster")
+
+	buckets := exoscalev1.BucketList{}
+	client := *o.k8sClient
+	log.V(1).Info("Listing buckets from cluster")
+	err := client.List(ctx, &buckets)
+	if err != nil {
+		return fmt.Errorf("cannot list buckets: %w", err)
+	}
+	o.bucketDetails = addOrgAndNamespaceToBucket(ctx, buckets)
+	return nil
+}
+
+func (o *ObjectStorage) saveToDatabase(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Creating a database connection")
+
+	log.V(1).Info("Opening database connection")
+	err := o.database.OpenConnection()
+	if err != nil {
+		return err
+	}
+	defer o.database.CloseConnection()
+
+	log.V(1).Info("Ensuring initial database configuration")
+	err = o.database.EnsureInitConfiguration(ctx)
+	if err != nil {
+		return err
+	}
+
+	log.V(1).Info("Saving buckets information usage to database")
+	for namespace, aggregatedBucket := range o.aggregatedBuckets {
+		err = o.database.EnsureBucketUsage(ctx, namespace, aggregatedBucket, o.billingDate)
+		if err != nil {
+			log.Info("WARNING: Buckets usage have not been saved to the database", "namespace", namespace, "error", err.Error())
+		}
+	}
+
+	return nil
+}
+
+func (o *ObjectStorage) getBillingDate(_ context.Context) error {
+	location, err := time.LoadLocation(exoscaleTimeZone)
+	if err != nil {
+		return fmt.Errorf("cannot initialize location from time zone %s: %w", location, err)
+	}
+	now := time.Now().In(location)
+	previousDay := now.Day() - 1
+	o.billingDate = time.Date(now.Year(), now.Month(), previousDay, exoscaleBillingHour, 0, 0, 0, now.Location())
+	return nil
+}
+
+func getAggregatedBuckets(ctx context.Context, sosBucketsUsage []oapi.SosBucketUsage, bucketDetails []BucketDetail) map[string]db.AggregatedBucket {
+	log := ctrl.LoggerFrom(ctx)
+	log.Info("Aggregating buckets by namespace")
+	aggregatedBuckets := make(map[string]db.AggregatedBucket)
+	for _, bucketDetail := range bucketDetails {
+		log.V(1).Info("Checking bucket", "bucket", bucketDetail.BucketName)
+		storageUsed := float64(0)
+		usageBucketExists := false
+
+		// Match managed kubernetes resource bucket to the exoscale usage bucket to retrieve the amount of storage used
+		for _, bucketUsage := range sosBucketsUsage {
+			if bucketDetail.BucketName == *bucketUsage.Name {
+				log.V(1).Info("Found exoscale bucket usage", "bucket", bucketUsage.Name, "bucket size", bucketUsage.Name)
+				storageUsed = float64(*bucketUsage.Size)
+				usageBucketExists = true
+				break
+			}
+		}
+
```

Comment:
Transforming a slice into a map doesn't require to loop over all the elements of the slice?

Reply:
I understand.
However, think of this problem the next time you have for-loops within for-loops. Remember Seiso? We have the same problem there with nested for-loops, probably even O(n3) and the performance impact is *huge*, it takes Seiso *MINUTES* to figure out if something can be deleted with just 20-40 Kubernetes resources :smile:

Reply:
I think you misunderstand me.
Here's what I mean with the map, the full code:
```go

func getAggregatedBuckets(ctx context.Context, sosBucketsUsage []oapi.SosBucketUsage, bucketDetails []BucketDetail) map[string]db.AggregatedBucket {
	log := ctrl.LoggerFrom(ctx)
	log.Info("Aggregating buckets by namespace")

	sosBucketsUsageMap := make(map[string]oapi.SosBucketUsage)
	for _, usage := range sosBucketsUsage {
		sosBucketsUsageMap[*usage.Name] = usage
	}

	aggregatedBuckets := make(map[string]db.AggregatedBucket)
	for _, bucketDetail := range bucketDetails {
		log.V(1).Info("Checking bucket", "bucket", bucketDetail.BucketName)
		storageUsed := float64(0)

		if bucketUsage, exists := sosBucketsUsageMap[bucketDetail.BucketName]; exists {
			log.V(1).Info("Found exoscale bucket usage", "bucket", bucketUsage.Name, "bucket size", bucketUsage.Name)
			storageUsed = float64(*bucketUsage.Size)
			aggregatedBucket := aggregatedBuckets[bucketDetail.Namespace]
			aggregatedBucket.Organization = bucketDetail.Organization
			aggregatedBucket.StorageUsed += storageUsed
			aggregatedBuckets[bucketDetail.Namespace] = aggregatedBucket
		} else {
			log.Info("Could not find any bucket on exoscale", "bucket", bucketDetail.BucketName)
		}
	}
	return aggregatedBuckets
}
```
The unit tests still pass without any change.

Reply:
Yup seems like a more cleaner version. Thanks for the optimization.

Reply:
@ccremer just curious, did you also run a go benchmark on that thing? You trade CPU cycles with memory allocation operations, so I'd expect it do be a bit slower for a small amount of buckets. The same effect can be observed with quicksort (O(logN)) vs insertion (O(N2)) sort for a low number of items...

Anyway that's just something that piqued my interest, not really important to the problem at hand :)
I agree that Chris' implementation will scale much better with many objects.

Reply:
I didn't want to get further distracted from my work by benchmarking this case.
I found this article, which says that after around 5-10 elements map lookups are faster: https://darkcoding.net/software/go-slice-search-vs-map-lookup/ (although it's from 2015, quite possibly things improved since then)

Reply:
@ccremer thanks for sharing, I tried to find such an article yesterday but couldn't. It's true that the hash used in maps is much faster and that is the approach to take when mathcing objects from 2 data structures. Also check out that the delay is only 0.002 ms per 100 items which translates to 2 ms for 100k items which is a bit irrelevant for small applications like this one though it is quite bad for 1mln + len slices.

Reply:
Sure, it's irrelevant here, but still nice to know about these kinds of problems, to avoid bad code smells.